File: ReSharper.FSharp/src/Psi.FSharp/src/FSharpProjectFileLanguageService.cs
using JetBrains.Platform.ProjectModel.FSharp;
using JetBrains.ProjectModel;
using JetBrains.ProjectModel.Resources;
using JetBrains.ReSharper.Plugins.FSharp.Common.CheckerService;
using JetBrains.ReSharper.Psi.FSharp.Parsing;
using JetBrains.ReSharper.Psi.Parsing;
using JetBrains.Text;
using JetBrains.UI.Icons;

namespace JetBrains.ReSharper.Psi.FSharp
{
  [ProjectFileType(typeof(FSharpProjectFileType))]
  public class FSharpProjectFileLanguageService : ProjectFileLanguageService
  {
    private readonly FSharpCheckerService myCheckerService;

    public FSharpProjectFileLanguageService(ProjectFileType projectFileType, FSharpCheckerService checkerService)
      : base(projectFileType)
    {
      myCheckerService = checkerService;
    }

    protected override PsiLanguageType PsiLanguageType => FSharpLanguage.Instance;
    public override IconId Icon => ProjectModelThemedIcons.Fsharp.Id;

    public override ILexerFactory GetMixedLexerFactory(ISolution solution, IBuffer buffer,
      IPsiSourceFile sourceFile = null)
    {
      return sourceFile != null
        ? new FSharpLexerFactory(sourceFile, myCheckerService.GetDefines(sourceFile))
        : FSharpLanguage.Instance.LanguageService().GetPrimaryLexerFactory();
    }

    public override IPsiSourceFileProperties GetPsiProperties(IProjectFile projectFile, IPsiSourceFile sourceFile,
      IsCompileService isCompileService)
    {
      return new FSharpPsiProperties(projectFile, sourceFile);
    }
  }
}

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/LanguageService/FSharpProjectFileLanguageService.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Psi.LanguageService

open System.Runtime.InteropServices
open JetBrains.ProjectModel
open JetBrains.ProjectModel.Resources
open JetBrains.ReSharper.Plugins.FSharp
open JetBrains.ReSharper.Plugins.FSharp.Checker
open JetBrains.ReSharper.Plugins.FSharp.ProjectModel.Scripts
open JetBrains.ReSharper.Plugins.FSharp.Psi
open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
open JetBrains.ReSharper.Plugins.FSharp.Util.FSharpMsBuildUtils
open JetBrains.ReSharper.Psi

[<ProjectFileType(typeof<FSharpProjectFileType>)>]
type FSharpProjectFileLanguageService(projectFileType, fsCheckerService: FSharpCheckerService,
        fsFileService: IFSharpFileService) =
    inherit ProjectFileLanguageService(projectFileType)

    override x.PsiLanguageType = FSharpLanguage.Instance :> _
    override x.Icon = ProjectModelThemedIcons.Fsharp.Id

    override x.GetMixedLexerFactory(_, _, [<Optional; DefaultParameterValue(null: IPsiSourceFile)>] sourceFile) =
        match sourceFile with
        | null -> FSharpLanguage.Instance.LanguageService().GetPrimaryLexerFactory()
        | _ ->

        let defines = fsCheckerService.FcsProjectProvider.GetParsingOptions(sourceFile).ConditionalCompilationDefines
        FSharpPreprocessedLexerFactory(defines) :> _

    override x.GetPsiProperties(projectFile, sourceFile, isCompileService) =
        let providesCodeModel =
            // todo: use items container instead
            isCompileService.IsCompile(projectFile, sourceFile) ||
            fsFileService.IsScriptLike(sourceFile) ||

            let buildAction = projectFile.Properties.GetBuildAction(sourceFile.PsiModule.TargetFrameworkId)
            buildAction = BuildActions.compileBefore || buildAction = BuildActions.compileAfter
        FSharpPsiProperties(projectFile, sourceFile, providesCodeModel) :> _


[<ProjectFileType(typeof<FSharpSignatureProjectFileType>)>]
type FSharpSignatureProjectFileLanguageService(projectFileType, fsCheckerService, fsFileService) =
    inherit FSharpProjectFileLanguageService(projectFileType, fsCheckerService, fsFileService)

    override x.PsiLanguageType = FSharpLanguage.Instance :> _


[<ProjectFileType(typeof<FSharpScriptProjectFileType>)>]
type FSharpScriptProjectFileLanguageService(projectFileType, fsCheckerService, fsFileService) =
    inherit FSharpProjectFileLanguageService(projectFileType, fsCheckerService, fsFileService)

    override x.PsiLanguageType = FSharpScriptLanguage.Instance :> _